File: CIT/julia/src/Generator.jl
module GeneratorCheck

using LinearAlgebra, SparseArrays
using ..Types: ModelParameters, validate

struct GeneratorResult
    grid::Vector{Float64}
    stationary_mass::Vector{Float64}
    response_direction::Vector{Float64}
    poisson_solution::Vector{Float64}
    susceptibility::Float64
    threshold::Float64
end

function build_generator(p::ModelParameters; grid_points::Int=801, domain::Float64=5.0)
    validate(p)
    grid_points >= 51 && isodd(grid_points) || throw(ArgumentError("grid_points must be odd and at least 51"))
    grid = collect(range(-domain, domain; length=grid_points))
    dx = grid[2] - grid[1]
    diffusion = 0.5 * p.diffusion^2
    Q = spzeros(Float64, grid_points, grid_points)
    for (i,u) in enumerate(grid)
        drift = -p.mean_reversion * u
        rate_plus = diffusion / dx^2 + max(drift, 0.0) / dx
        rate_minus = diffusion / dx^2 + max(-drift, 0.0) / dx
        Q[i, i < grid_points ? i+1 : i-1] += rate_plus
        Q[i, i > 1 ? i-1 : i+1] += rate_minus
        if abs(u) > p.tolerance
            target = sign(u) * p.reset_fraction * p.tolerance
            j = argmin(abs.(grid .- target))
            j != i && (Q[i,j] += p.active_intensity)
        end
        Q[i,i] = -sum(Q[i,:])
    end
    grid, Q
end

function stationary_mass(Q::SparseMatrixCSC{Float64,Int})
    n = size(Q,1)
    A = Matrix(transpose(Q)); b = zeros(n)
    A[end,:] .= 1.0; b[end] = 1.0
    mass = max.(A \ b, 0.0)
    mass ./ sum(mass)
end

function translation_direction(grid, mass)
    dx = grid[2] - grid[1]
    direction = similar(mass)
    direction[2:end-1] .= -(mass[3:end] .- mass[1:end-2]) ./ (2 * dx)
    direction[1] = -(mass[2]-mass[1]) / dx
    direction[end] = -(mass[end]-mass[end-1]) / dx
    direction .-= mass .* sum(direction)
    normalization = dot(grid, direction)
    abs(normalization) > 1e-12 || error("response direction normalization failed")
    direction ./ normalization
end

function solve_poisson(Q, direction)
    n = size(Q,1)
    A = Matrix(-transpose(Q)); b = copy(direction)
    A[end,:] .= 1.0; b[end] = 0.0
    A \ b
end

function generator_cross_check(p::ModelParameters; grid_points::Int=801, domain::Float64=5.0)
    grid, Q = build_generator(p; grid_points, domain)
    mass = stationary_mass(Q)
    direction = translation_direction(grid, mass)
    poisson = solve_poisson(Q, direction)
    susceptibility = dot(grid, poisson)
    GeneratorResult(grid, mass, direction, poisson, susceptibility, inv(susceptibility))
end

end

File: CIT/python/src/cit/generator.py
from __future__ import annotations

from dataclasses import dataclass
import numpy as np
from scipy.sparse import csr_matrix, lil_matrix
from scipy.sparse.linalg import spsolve
from .model import ModelParameters

@dataclass(frozen=True)
class GeneratorResult:
    grid: np.ndarray
    stationary_mass: np.ndarray
    response_direction: np.ndarray
    poisson_solution: np.ndarray
    susceptibility: float
    threshold: float

def build_generator(model: ModelParameters, grid_points: int = 801, domain: float = 5.0) -> tuple[np.ndarray, csr_matrix]:
    model.validate()
    if grid_points < 51 or grid_points % 2 == 0:
        raise ValueError("grid_points must be odd and at least 51")
    grid = np.linspace(-domain, domain, grid_points)
    dx = grid[1] - grid[0]
    diffusion = 0.5 * model.diffusion**2
    generator = lil_matrix((grid_points, grid_points))
    for i, u in enumerate(grid):
        drift = -model.mean_reversion * u
        rate_plus = diffusion / dx**2 + max(drift, 0.0) / dx
        rate_minus = diffusion / dx**2 + max(-drift, 0.0) / dx
        generator[i, i + 1 if i < grid_points - 1 else i - 1] += rate_plus
        generator[i, i - 1 if i > 0 else i + 1] += rate_minus
        if abs(u) > model.tolerance:
            target = np.sign(u) * model.reset_fraction * model.tolerance
            target_index = int(np.argmin(np.abs(grid - target)))
            if target_index != i:
                generator[i, target_index] += model.active_intensity
        generator[i, i] = -generator[i].sum()
    return grid, generator.tocsr()

def generator_cross_check(model: ModelParameters, grid_points: int = 801, domain: float = 5.0) -> GeneratorResult:
    grid, generator = build_generator(model, grid_points, domain)
    n = generator.shape[0]
    system = generator.T.tolil(); rhs = np.zeros(n)
    system[-1, :] = 1.0; rhs[-1] = 1.0
    mass = np.maximum(spsolve(system.tocsr(), rhs), 0.0); mass /= mass.sum()
    dx = grid[1] - grid[0]
    direction = np.empty_like(mass)
    direction[1:-1] = -(mass[2:] - mass[:-2]) / (2 * dx)
    direction[0] = -(mass[1] - mass[0]) / dx
    direction[-1] = -(mass[-1] - mass[-2]) / dx
    direction -= mass * direction.sum(); direction /= float(grid @ direction)
    system = (-generator.T).tolil(); rhs = direction.copy()
    system[-1, :] = 1.0; rhs[-1] = 0.0
    poisson = spsolve(system.tocsr(), rhs)
    susceptibility = float(grid @ poisson)
    return GeneratorResult(grid, mass, direction, poisson, susceptibility, 1.0 / susceptibility)
